working/control-flow-collections/feature-specification.md

```diff
@@ -0,0 +1,913 @@
+single element it contains.
+
+### Scoping
+
+Both styles of `for` element may introduce a local variable, as in:
+
+```dart
+[
+  for (var i = 1; i < 4; i++) i,
+  for (var i in [1, 2, 3]) i
+]
+```
+
+If a `for` element declares a variable, then a new namespace is created where
+that variable is defined. The body of the `for` element is resolved and
+evaluated in that namespace. The variable goes out of scope at the end of the
+element's body.
+
+Each iteration of the loop binds a new fresh variable:
+
+```dart
+var closures = [for (var i = 1; i < 4; i++) () => i];
+for (var closure in closures) print(closure());
+// Prints "1", "2", "3".
+```
+
+### Static errors
+
+The static semantics of collection `if` and `for` mostly follow their statement
+analogues.
+
+If is a static error when:
+
+*   The collection is a list and the type of any of the body elements may not be
+    assigned to the list's element type.
+
+    ```dart
+    <int>[if (true) "not int"] // Error.
+    ```
+
+*   The collection is a map and the key type of any of the body elements may not
+    be assigned to the map's key type.
+
+    ```dart
+    <int, int>{if (true) "not int": 1} // Error.
+    ```
+
+*   The collection is a map and the value type of any of the body elements may
+    not be assigned to the map's value type.
+
+    ```dart
+    <int, int>{if (true) 1: "not int"} // Error.
+    ```
+
+*   The type of the condition expression in an `if` element may not be assigned
+    to `bool`.
+
+    ```dart
+    [if ("not bool") 1] // Error.
+    ```
+
+*   The type of the iterator expression in a `for-in` element may not be
+    assigned to `Iterable<T>` for some type `T`. Otherwise, the *iterable type*
+    of the iterator is `T`.
+
+    ```dart
+    [for (var i in "not iterable") i] // Error.
+    ```
+
+*   The iterable type of the iterator in a `for-in` element may not be assigned
+    to the `for-in` variable's type.
+
+    ```dart
+    [for (int i in ["not", "int"]) i] // Error.
+    ```
+
+*   The type of the condition expression (the second clause) in a C-style `for`
+    element may not be assigned to `bool`.
+
+    ```dart
+    [for (; "not bool";) 1] // Error.
+    ```
+
+### Type inference
+
+Inference propagates upwards and downwards like you would expect. For the most
+part, inference flows "through" the `if` and `for` into the body element(s).
+
+*   If a list literal has a downwards inference type of `List<T>` for some `T`,
+    then the downwards inference context type of the body elements is `T`.
+
+    Thus:
+
+    ```dart
+    List<List<String>> i = [
+      if (true) [],
+      if (false) [] else [],
+      for (var i = 0; i < 1; i++) []
+    ];
+    ```
+
+    Produces a `List<List<String>>` containing three empty `List<String>`.
+
+*   The upwards inference element type of an `if` list element without an `else`
+    is the type of the "then" element.
+
+*   The upwards inference element type of an `if-else` list element is the least
```

Comment:
Does it ever matter that you find the LUB of these two first, before finding the LUB of that and the other elements?
If so, maybe it would be nicer to treat the if/else as introducing two elements in the following element type LUB.

Reply:
I thought about this a little. As far as I know, LUB is associative so it shouldn't matter, but I wouldn't be surprised if there were subtleties I'm missing.

I defaulted to defining this locally instead of flattening because the latter felt a little weird when you consider that the elements of a list may have access to a variable that is out of scope at the top level of the list:

```dart
[for (var i in stuff) if (cond) i.something else i.other]
```

I wasn't sure if flattening the then and else parts would make things weird given that `i` is only meaningful inside the `for`.